sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala

```diff
@@ -202,21 +214,30 @@ private[csv] class UnivocityParser(
       }
       numMalformedRecords += 1
       None
-    } else if (options.failFast && schema.length != tokens.length) {
+    } else if (options.failFast && inputSchema.length != tokens.length) {
       throw new RuntimeException(s"Malformed line in FAILFAST mode: " +
         s"${tokens.mkString(options.delimiter.toString)}")
     } else {
-      val checkedTokens = if (options.permissive && schema.length > tokens.length) {
-        tokens ++ new Array[String](schema.length - tokens.length)
-      } else if (options.permissive && schema.length < tokens.length) {
-        tokens.take(schema.length)
+      if (options.permissive && shouldHandleCorruptRecord) {
+        row.setNullAt(corruptIndex)
+      }
+      val checkedTokens = if (options.permissive && inputSchema.length > tokens.length) {
+        tokens ++ new Array[String](inputSchema.length - tokens.length)
+      } else if (options.permissive && inputSchema.length < tokens.length) {
+        tokens.take(inputSchema.length)
       } else {
         tokens
       }
 
       try {
         Some(convert(checkedTokens))
       } catch {
+        // We only catch exceptions about malformed values here and pass over other exceptions
+        // (e.g., SparkException about unsupported types).
+        case _: NumberFormatException | _: IllegalArgumentException
+            if options.permissive && shouldHandleCorruptRecord =>
+          row(corruptIndex) = UTF8String.fromString(tokens.mkString(options.delimiter.toString))
```

Comment:
I added `CSVUtils.verifySchema` in the head of `buildReader`, then modified the code above to handle all the `NonFatal` exceptions with a permissive mode in `parseWithMode`.